### PR TITLE
WIP - Empty group values

### PR DESCRIPTION
**Pull request by febbraiod**
Hoist P/R Checklist
-------------------

**Pull request authors:** Review and check off the below. Items that do not apply can also be
checked off to indicate they have been considered. If unclear if a step is relevant, please leave
unchecked and note in comments.

- [ ] Caught up with `develop` branch as of last change.
- [ ] Added CHANGELOG entry, or determined not required.
- [ ] Reviewed for breaking changes, added `breaking-change` label + CHANGELOG if so.
- [ ] Updated doc comments / prop-types, or determined not required.
- [ ] Reviewed and tested on Mobile, or determined not required.
- [ ] Created Toolbox branch / PR, or determined not required.

**If your change is still a WIP**, please use the "Create draft pull request" option in the split
button below to indicate it is not ready yet for a final review.

> **Pull request reviewers:** when merging this P/R, please consider using a *squash commit* to
> collapse multiple intermediate commits into a single commit representing the overall feature
> change. This helps keep the commit log clean and easy to scan across releases. PRs containing a
> single commit should be *rebased* when possible.

### CODE REVIEW

File: cmp/grid/columns/Column.js [hunk 1]
```diff
@@ -385,8 +391,11 @@ export class Column {
             setElementRenderer(class extends Component {
                 render() {
                     const agParams = this.props,
-                        {value, data: record} = agParams;
-                    return elementRenderer(value, {record, column: me, gridModel, agParams});
+                        {value, data: record} = agParams,
+                        isGrouped = gridModel.groupBy.includes(colId);
+
+
+                    return isGrouped && emptyGroupedByValue && isEmpty(value) ?  elementRenderer(emptyGroupedByValue, {record, column: me, gridModel, agParams}) : elementRenderer(value, {record, column: me, gridModel, agParams});
```
Comment (febbraiod) [inline cmp/grid/columns/Column.js]: Not working yet

File: cmp/grid/columns/Column.js [hunk 2]
```diff
@@ -453,6 +462,24 @@ export class Column {
         if (isArray(fieldPath)) return get(record.data, fieldPath);
         return record.data[fieldPath];
     };
+
+    generateEmptyGroupGetValueFn(getValueFn) {
+        return (params) => {
+            const v = getValueFn(params),
+                isGrouped = params.gridModel.groupBy.includes(this.colId);
+            return isGrouped && isEmpty(v) ? this.emptyGroupedValue : v;
```
Comment (febbraiod) [inline cmp/grid/columns/Column.js]: IsEmpty(v) or v == null?

**Comment by febbraiod** [comment]
Pausing work on this at Lee's request for Anselm and him to review.

**Comment by lbwexler** [comment]
Thanks Don!

John -- Can you look at this one and give some thoughts?  

Is the cost-benefit worth it on this one?

**Comment by haynesjm42** [comment]
How often have we run into this specific use-case where you can't just deal with this in `processRawData` because you absolutely can't have your nil/empty field values be set to a placeholder value, and don't want to see that placeholder in normal columns, only in group rows?

What the app would need to do to handle this specific case themselves really isn't that bad, and we could just add an example of dealing with unbalanced groups to toolbox.
- Add a `getValueFn` on your column which returns your placeholder value when the field value is empty - this will be used for grouping _and_ for rendering in the group cell for the autoGroupColumn
- Add a `renderer` to your column which checks if the value is equal to your placeholder value, and return null or an empty string in that case, otherwise just return the value

This is pretty consistent with what ag-grid recommends for dealing with unbalanced groups: https://www.ag-grid.com/javascript-grid-grouping/#unbalanced-groups

That being said, I think Don's solution is creative and if we can handle this case better than ag-grid, then great - but I do get concerned about adding additional complexity to our Column, and am especially concerned about adding any additional processing to our renderers - we should definitely make sure we are only using the more complex renderers when we need them.

Another potential solution to this would be for us to reverse how we are handling the rendering - we could convert nil values to an empty string in the default `getValueFn` and then also install an innerRenderer for the autoGroupColumn which converts the empty value into the `emptyGroupedValue` in the GridModel

**Comment by lbwexler** [comment]
Thanks for the hard work here Don.

I am going to agree with  John's Assesment here -- I think the risk is that the cure here is worse than the disease.  It complicates a pretty complicated story already.  Let's let apps deal with this by pre-processing their nulls, either before presenting to the grid, or before loading the store its

**Pull request closed without merging**